core/iwasm/interpreter/wasm_loader.c

```diff
@@ -4623,6 +4623,14 @@ wasm_loader_ctx_init(WASMFunction *func)
               loader_malloc(loader_ctx->const_buf_size, NULL, 0)))
         goto fail;
 
+    if ((func->local_cell_num >= INT16_MAX)
+        || (func->param_cell_num >= INT16_MAX - func->local_cell_num)) {
```

Comment:
Integer overflow of `param_cell_num/ret_cell_num` is checked in `load_type_section`, L605.
Seems that we had better also check integer overflow of param_cell_num in `load_function_section`, L1956:
```C
if (wasm_get_cell_num(func->local_types, func->local_count)  >= INT16_MAX)
```
And here, had better convert INT16_MAX to uint32?
```C
if ((func->local_cell_num >= INT16_MAX)
        || (func->param_cell_num >= (uint32)INT16_MAX - func->local_cell_num)) {
```

Reply:
> Integer overflow of param_cell_num/ret_cell_num is checked in load_type_section, L605.

Yes, in L605 we check whether the cell num bigger than `UINT16_MAX (65535)`, but here we need to check `local_cell_num + param_cell_num` < `INT16_MAX(32767)`

> Seems that we had better also check integer overflow of param_cell_num in load_function_section, L1956:

I thinks you mean `local_cell_num` rather than `param_cell_num`?

Reply:
Seems we don't need to check in L1956, the `param_cell_num` and `local_cell_num` are both defined as `uint16`, so they won't be large than UINT16_MAX. 

Here we check whether `param_cell_num + local_cell_num < INT16_MAX`, this limit is from fast interpreter rather than wasm spec, because the upper bound of dynamic space is 32767

Reply:
> 



> > Integer overflow of param_cell_num/ret_cell_num is checked in load_type_section, L605.
> 
> Yes, in L605 we check whether the cell num bigger than `UINT16_MAX (65535)`, but here we need to check `local_cell_num + param_cell_num` < `INT16_MAX(32767)`

Yes they are different.

> 
> > Seems that we had better also check integer overflow of param_cell_num in load_function_section, L1956:
> 
> I thinks you mean `local_cell_num` rather than `param_cell_num`?

Yes, should be local_cell_num.

Reply:
Done.

I didn't add check in L1956 since `local_cell_num` is already uint16, it won't be large than UINT16_MAX

Reply:
> Seems we don't need to check in L1956, the `param_cell_num` and `local_cell_num` are both defined as `uint16`, so they won't be large than UINT16_MAX.
> 

L1956 converts uint32 into uint16, note that return type of wasm_get_cell_num() is uint32, but local_cell_num type is uint16, had better check the integer overflow like L605

> Here we check whether `param_cell_num + local_cell_num < INT16_MAX`, this limit is from fast interpreter rather than wasm spec, because the upper bound of dynamic space is 32767

Yes, here we add extra check for `param_cell_num + local_cell_num < INT16_MAX`

Reply:
> L1956 converts uint32 into uint16, note that return type of wasm_get_cell_num() is uint32, but local_cell_num type is uint16, had better check the integer overflow like L605

Got it, thanks!
I'll check it.

Reply:
Added check in L1956.

> And here, had better convert INT16_MAX to uint32?

INT16_MAX should be converted to `int32`, otherwise the final result will be treated as `uint32`, which may cause integer underflow

Reply:
OK